api/v1alpha4/cluster_types.go

```diff
@@ -215,6 +216,43 @@ func (c *Cluster) SetConditions(conditions Conditions) {
 	c.Status.Conditions = conditions
 }
 
+func (c *Cluster) IPFamily() (ClusterIPFamily, error) {
```

Comment:
I see... if we want to continue to pursue the auto-detection route, then I think this may benefit from a table test of sorts:

When service CIDR and pod CIDR are both specified and match there is no ambiguity.

I think these cases are also pretty safe to determine:

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| (empty) | ipv4 | ipv4 |
| (empty) | ipv6 | ipv6 |
| (empty) | ipv4,ipv6 | dualstack |
| ipv4,ipv6 | (empty) | dualstack |
| ipv4 | ipv4,ipv6 | dualstack |
| ipv6 | ipv4,ipv6 | dualstack |


When they are both specified an IP Family within service CIDR list is not in the pod CIDR list, then it should be invalid, for example:

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| ipv4 | ipv6 | invalid | 
| ipv4,ipv6 | ipv6 | invalid | 
| ipv6 | ipv4 | invalid | 
| ipv4,ipv6 | ipv6 | invalid | 

It is just these cases that I believe are ambiguous, as they could be dualstack. Perhaps in these cases we could choose a default as shown?

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| (empty) | (empty) | ipv4 | could be dualstack
| ipv4 | (empty) | ipv4 | could be dualstack
| ipv6 | (empty) | ipv6 | could be dualstack

Reply:
This is what we have in CAPZ currently to support single stack IPv6 on Azure: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/cluster-template-ipv6.yaml

this is the workaround we had to apply to get it working: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/cluster-template-ipv6.yaml#L147-L150

+1 to auto-detection, it'd be great if we can avoid requiring changes for providers

Reply:
Also what we've done so far is check the vnet CIDR to determine if it's using ipv6: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/scope/cluster.go#L301

Reply:
Oh, thanks for pointing that out. I see there's a bunch of utility functions here that may be useful at least to reference: https://github.com/kubernetes/utils/blob/master/net/net.go

Reply:
OK, updated this logic to match the table above with tests that should make it easier to review the various categories of cases here. Let me know if you have any additional thoughts on this auto-detection now with these changes.

Reply:
@mcwumby looks great to me! might be we should add a "configuring network" page in CAPI book under tasks, but this could be also in a separated PR.

@vincepri @CecileRobertMichon given that we are now identifying a set of invalid network configurations, should we error out in the validation web hook for the Cluster object?

Reply:
FWIW, some discussion about this on sig-networking: https://groups.google.com/g/kubernetes-sig-network/c/S_64nw4BqoM/m/w64pw2ejAwAJ